cumulusci/cli/ui.py

```diff
@@ -133,3 +133,137 @@ def _dim_value(self, val):
+
+    if format == "table":
+        help_message = "Type help(query) to learn about other return formats or assigning the result."
+        print(_soql_table(results, truncated))
+        print()
+        print(help_message)
+        rc = None
+    elif format == "obj":
+        rc = results
+    elif format == "pprint":
+        from pprint import pprint
+
+        pprint(results)
+        rc = None
+    elif format == "json":
+        from json import dumps
+
+        rc = dumps(results, indent=2)
+    else:
+        raise TypeError(f"Unknown format `{format}`")
+
+    return rc
+
+
+def _soql_table(results, truncated):
+    if results:
+        if truncated:
+            assert results[-1] == truncated
+            first_row = results[0]
+            fake_row = {k: "" for k, v in first_row.items()}
+            first_column = list(first_row)[0]
+            fake_row[first_column] = truncated
+
+            results[-1] = fake_row
+
+        headings = list(results[0].keys())
+        return str(
+            CliTable(
+                [headings] + [list(map(str, r.values())) for r in results],
+                wrap_cols=headings,
+            )
+        )
+    else:
+        return str(CliTable([["No results"]]))
+
+
+def pretty_describe(sf, sobj_name, detailed, format):
+    """Describe an sobject.
+
+    Arguments:
+
+        sobj_name - sobject name to describe. e.g. "Account", "Contact"
+        detailed - set to `True` to get detailed information about object
+        format -- one of these values:
+           - "obj" -- ordinary Python objects (default)
+           - "pprint" -- string in easily readable Python dict shape
+        """
+    from pprint import pprint
+
+    data = getattr(sf, sobj_name).describe()
+
+    if detailed:
+        rc = data
+    else:
+        rc = dict(_summarize(field) for field in data["fields"])
+
+    if format == "pprint":
+        pprint(rc)
+    elif format == "obj":
+        return rc
+    else:
+        raise TypeError(f"Unknown format {format}")
+
+
+def _summarize(field):
+    if field["referenceTo"]:
+        allowed = field["referenceTo"]
+    elif field["picklistValues"]:
+        allowed = [value["value"] for value in field["picklistValues"]]
+    else:
+        allowed = field["type"]
+    return (field["name"], allowed)
+
+
+def repl_helpers(sf):
+    # these don't use functools.partial because it has a weird __repr__
+    def query(soql_text, include_deleted=False, format="table", max_rows=100):
+        return pretty_soql_query(
+            sf, soql_text, include_deleted=False, format=format, max_rows=max_rows
+        )
+
+    query.__doc__ = pretty_soql_query.__doc__.replace("pretty_soql_query", "query")
+
+    def describe(obj_name, detailed=False, format="pprint"):
+        return pretty_describe(sf, obj_name, detailed=detailed, format=format)
+
+    describe.__doc__ = pretty_describe.__doc__.replace("pretty_describe", "describe")
+
+    return locals()
```

Comment:
This seems like a strange and unnecessarily clever way of defining these helpers and connecting them to a simple-salesforce instance. Is there a reason not to do something like this?

    class SalesforceAPIAdapter:
        def __init__(self, sf):
            self.sf = sf

        def pretty_soql_query(self, ...):
            ...

        def pretty_describe(self, ...):
            ...

and then use it like this:

    sf_adapter = SalesforceAPIAdapter(sf)
    globals = {
        ...
        "query": sf_adapter.pretty_soql_query,
        "describe": sf_adapter.pretty_describe,
    }

Reply:
Okay yeah, it just evolved from a partial. I've implemented a class similar to what you described: 51ab2f4d